pkg/role/datastore.go

```diff
@@ -0,0 +1,162 @@
+package role
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/libopenstorage/openstorage/api"
+	"github.com/portworx/kvdb"
+)
+
+// Datastore provides APIs for storing and fetching SDK Roles
+// from a persistent datastore
+type Datastore interface {
+
+	// Exists returns true if the role exists in the datastore
+	Exists(roleName string) (bool, error)
+
+	// Update updates the role in the datastore. It will return an error
+	// if the role does not exist in the datastore
+	Update(roleName string, sdkRole *api.SdkRole) error
+
+	// Put puts the role in the datastore. It will overwrite the
+	// role object if it already exists
+	Put(roleName string, sdkRole *api.SdkRole) error
+
+	// Get gets the role object from the datastore for the given name
+	Get(roleName string) (*api.SdkRole, error)
+
+	// Create creates the role in the datastore. It will return an error
+	// if the role with the name already exists
+	Create(roleName string, sdkRole *api.SdkRole) error
+
+	// EnumerateRoleNames enumerates all the roles from the datastore and returns their names
+	EnumerateRoleNames() ([]string, error)
+
+	// Delete deletes the role from the datastore
+	Delete(roleName string) error
+}
+
+var (
+	// ErrRoleNotFound is returned when role is not found in datastore
+	ErrRoleNotFound = errors.New("role not found in datastore")
+	// ErrRoleExists is returned when a role with the same name exists
+	ErrRoleExists = errors.New("role exists with the same name")
+)
+
+const (
+	rolePrefix = "cluster/roles"
+)
+
+// Simple function which creates key for Kvdb
+func prefixWithName(name string) string {
+	return rolePrefix + "/" + name
+}
+
+type kvRoleDatastore struct {
```

Comment:
All these methods are not part of the gRPC interface as they used to be before which included the `ctx`. If possible, we should try to maintain the receivers to maintain the ctx variable which may have a lot of information about the request like the request id (in a future release).

Reply:
Do you mean we should pass in the Context to all the APIs in this Datastore interface? 
I thought that the layer which talks to a persistence store should not worry about who is requesting it and rely on their callers to do the proper context checks.